File: CodeTraining/LeetCode/easy/linklist/Code_234.java
package LeetCode.easy.linklist;

public class Code_234 {
    private class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    private ListNode frontPointer;

    public boolean isPalindrome(ListNode head) {
        frontPointer = head;
        return recursivelyCheck(head);
    }

    private boolean recursivelyCheck(ListNode head) {
        if(head != null){
            if (!recursivelyCheck(head.next)) return false;
            if(head.val != frontPointer.val) return false;
            frontPointer = frontPointer.next;
        }
        return true;
    }
}

File: CodeTraining/LeetCode/easy/linklist/code_234.scala
package LeetCode.easy.linklist

import javax.management.ListenerNotFoundException

object code_234 {

  private class ListNode(var _x: Int = 0) {
    var next: ListNode = null
    var x: Int = _x
  }

  private var frontpointer = new ListNode()

  def isPalindrome(head: ListNode): Boolean = {
    frontpointer = head
    recursivelyCheck(head)
  }

  def recursivelyCheck(node: ListNode): Boolean ={
    if(node != null){
      if(!recursivelyCheck(node.next)) return false
      if(node.x != frontpointer.x) return false
      frontpointer = frontpointer.next
    }
    true
  }

}
